tools/android/packaging/xbmc/src/org/xbmc/kodi/XBMCJsonRPC.java.in

```diff
@@ -0,0 +1,647 @@
+        e.printStackTrace();
+        return null;
+      }
+    } catch (Exception e)
+    {
+      Log.e(TAG, "Failed to read JSON");
+      e.printStackTrace();
+      return null;
+    }
+  }
+
+  public JSONObject request_object(String jsonRequest)
+  {
+    try
+    {
+      JSONObject resp = new JSONObject(request_string(jsonRequest));
+      return resp;
+    } catch (Exception e)
+    {
+      Log.e(TAG, "Failed to parse JSON");
+      e.printStackTrace();
+      return null;
+    }
+  }
+
+  public JSONArray request_array(String jsonRequest)
+  {
+    try
+    {
+      JSONArray resp = new JSONArray(request_string(jsonRequest));
+      return resp;
+    } catch (Exception e)
+    {
+      Log.e(TAG, "Failed to parse JSON");
+      e.printStackTrace();
+      return null;
+    }
+  }
+
+  public Bitmap getBitmap(String src)
+  {
+    if (src.isEmpty())
+      return null;
+
+    try
+    {
+      JSONObject req = request_object("{\"jsonrpc\": \"2.0\", \"method\": \"Files.PrepareDownload\", \"params\": { \"path\": \""
+          + src + "\"}, \"id\": \"1\"}");
+      if (req == null)
+        return null;
+
+      JSONObject result = req.getJSONObject("result");
+      String surl = result.getJSONObject("details").getString("path");
+
+      URL url = new URL(m_jsonURL + "/" + surl);
+      HttpURLConnection connection = (HttpURLConnection) url.openConnection();
+      connection.setDoInput(true);
+      connection.connect();
+      InputStream input = connection.getInputStream();
+      Bitmap myBitmap = BitmapFactory.decodeStream(input);
+      return myBitmap;
+    } catch (Exception e)
+    {
+      e.printStackTrace();
+      return null;
+    }
+  }
+
+  public String getBitmapUrl(String src)
+  {
+    if (src.isEmpty())
+      return null;
+
+    try
+    {
+      JSONObject req = request_object("{\"jsonrpc\": \"2.0\", \"method\": \"Files.PrepareDownload\", \"params\": { \"path\": \""
+          + src + "\"}, \"id\": \"1\"}");
+      if (req == null)
+        return null;
+
+      JSONObject result = req.getJSONObject("result");
+      String surl = result.getJSONObject("details").getString("path");
+
+      return (m_jsonURL + "/" + surl);
+    } catch (Exception e)
+    {
+      e.printStackTrace();
+      return null;
+    }
+  }
+
+  public Cursor search(String query)
+  {
+      String[] menuCols = new String[] {
+              BaseColumns._ID,
+              COLUMN_TITLE,
+              COLUMN_TAGLINE,
+              COLUMN_THUMB,
+              COLUMN_FANART,
+      };
+      MatrixCursor mc = new MatrixCursor(menuCols);
+
+      try
+      {
+        JSONObject req = request_object(String.format(SEARCH_MOVIES_JSON, /*"\"operator\": \"contains\", \"field\": \"title\", \"value\": \"" + query + "\"", limit));*/
+        "\"or\": [" +
+        "{\"operator\": \"contains\", \"field\": \"title\", \"value\": \"" + query + "\"}," +
+        "{\"operator\": \"contains\", \"field\": \"set\", \"value\": \"" + query + "\"}," +
```

Comment:
I don't think I can answer that, as I only cherrypicked from @koying and tried to improve some parts. Pretty new to doing android stuff, so if you can point me in the right directions to improve it even more, that would be appreciated (not sure how much time I will have in the next months)